snakemake/executors.py

```diff
@@ -1596,7 +1596,13 @@ def __init__(
         max_status_checks_per_second=1,
     ):
 
-        self.workflow_sources = workflow.get_sources()
+        self.workflow_sources = list(workflow.get_sources())
+        for wfs in self.workflow_sources.copy():
+            if os.path.isdir(wfs):
+                for (dirpath, dirnames, filenames) in os.walk(wfs):
+                    self.workflow_sources.extend([os.path.join(dirpath, f) for f in filenames])
+                self.workflow_sources.remove(wfs)
+
```

Comment:
Please create self.workflow_sources in one pass, instead of deleting items (which is expensive, O(n)). You can have a generator expression in line 1599.

Reply:
ok I think I know what you mean.